### PR TITLE
[App] Add cloud platform exception

### PR DESCRIPTION
**Pull request by adam-lightning**
## What does this PR do?

We add a new `LightningPlatformException` as to indicate possible error events happening on the platform the Lightning App is on.
We currently consider (but future use is not limited to) the following cases:
- User tries to create new default Work and exceeds the allowed quota (10).
- User does not have enough credits and tries to create non-default Work.

This Exception is used on the launcher side, so here's just a definition of that exception.

This way if that's the user's intent, they can react accordingly to the situation.

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda

### CODE REVIEW

File: src/lightning_app/utilities/exceptions.py [hunk 1]
```diff
@@ -41,6 +41,10 @@ class LightningWorkException(Exception):
     """Exception used to inform users of misuse with LightningWork."""
 
 
+class LightningPlatformException(Exception):
+    """Exception used to inform users of issues related to platform the LightningApp is running on."""
+
```
Comment (awaelchli) [inline src/lightning_app/utilities/exceptions.py:46]: Can you point us to where this is being raised?

Comment (adam-lightning) [reply src/lightning_app/utilities/exceptions.py:46]: Directly in the lightning_launcher ([here](https://github.com/gridai/grid/pull/7789/files)) while creating a work resource. So it's not raised by framework - but for framework from platform.

Comment (adam-lightning) [reply src/lightning_app/utilities/exceptions.py]: @awaelchli Do you think that's now ok? Or you mean comment like 
```
# This is raised by platform for the framework to signal platform-dependent problems.
class Light...
```

File: src/lightning_app/utilities/exceptions.py [hunk 2]
```diff
@@ -41,6 +41,13 @@ class LightningWorkException(Exception):
     """Exception used to inform users of misuse with LightningWork."""
 
 
+class LightningPlatformException(Exception):  # pragma: no cover
+    """Exception used to inform users of issues related to platform the LightningApp is running on.
+
+    It's raised for framework by a platform LightningApp is running on.
```
Comment (awaelchli) [inline src/lightning_app/utilities/exceptions.py:47]:
```suggestion
    It gets raised by the Lightning Launcher on the platform side when the app is running in the cloud, and is useful
    when framework or user code needs to catch exceptions specific to the platform, e.g., when resources exceed
    quotas.
```

File: src/lightning_app/utilities/exceptions.py (hunk 1 above)
Comment (awaelchli) [reply src/lightning_app/utilities/exceptions.py]: Yes that's what I had in mind. 
I posted a suggestion how to improve the sentence.

**Review by awaelchli — APPROVED** [review]

**Review by hhsecond — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by otaj — APPROVED** [review]

File: src/lightning_app/utilities/exceptions.py [hunk 3]
```diff
@@ -41,6 +41,14 @@ class LightningWorkException(Exception):
     """Exception used to inform users of misuse with LightningWork."""
 
 
+class LightningPlatformException(Exception):  # pragma: no cover
```
Comment (Borda) [inline src/lightning_app/utilities/exceptions.py:44]: Where is this used?

**Review by Borda — APPROVED** [review]

File: src/lightning_app/utilities/exceptions.py (hunk 3 above)
Comment (adam-lightning) [reply src/lightning_app/utilities/exceptions.py:44]: @Borda On the side of the launcher (platform) 
The idea is to raise it if something happens on the platform that the framework should (could) react to (e.g. cannot start work due to out of credits)

**Pull request merged**